### PR TITLE
gcc: fix missing atomic ops errors on risc-v

### PR DESCRIPTION
**Pull request by fgaz**
## Description of changes

Fixes errors like:

    ld: [...]: undefined reference to `__atomic_exchange_1'

* Apply patch from upstream that uses libatomic for missing ops: https://gcc.gnu.org/git?p=gcc.git;a=commit;h=f797260adaf52bee0ec0e16190bbefbe1bfc3692 A few conflicts were resolved, so the patch is included as a file.
* Remove workarounds from
  * aws-sdk-cpp
  * libjxl
  * numactl
  * mariadb
  * zstd

Fixes https://github.com/NixOS/nixpkgs/pull/258614#issuecomment-1773235978

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
  - [x] riscv64-linux
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
  * [x] cmakeMinimal (was broken by this)
  * [x] aws-sdk-cpp
  * [ ] libjxl
  * [x] numactl
  * [ ] mariadb
  * [x] zstd
  * [ ] llvm and clang 17
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by ghost — COMMENTED** [review]
I'm not excited about carrying a gcc patch simply in order to eliminate `LDFLAGS` from a few packages.

It looks like this patch didn't make it into gcc 13.2, so if there is a workaround (which there is) we should be using the workaround instead until this patch shows up in a release.  Generally we ship merged-but-unreleased patches only when there is no alternative available.

File: pkgs/development/compilers/gcc/patches/default.nix [hunk 1]
```diff
@@ -151,6 +151,15 @@ in
 }.${majorVersion} or [])
 
 
+### RISC-V
+
+
+# https://gcc.gnu.org/git?p=gcc.git;a=commit;h=f797260adaf52bee0ec0e16190bbefbe1bfc3692
```
[the remark this answers is not in the document]
Comment (fgaz) [reply pkgs/development/compilers/gcc/patches/default.nix:157]: None yet as far as I can see

**Comment by fgaz** [comment]
> I'm not excited about carrying a gcc patch simply in order to eliminate LDFLAGS from a few packages.

I suspect that there are way more broken packages that we don't know about, since nobody builds the whole nixpkgs on risc-v.

For now I just discovered cmake and llvm 17, so I can certainly apply the workaround to those too.

Up to you.

**Comment by fgaz** [comment]
...and of course it's a mass rebuild because `optionalString false` returns an empty string, which is different from an empty LDFLAGS. If we proceed with this I guess I'll have to split the workaround removal.

File: pkgs/development/compilers/gcc/patches/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (fgaz) [reply pkgs/development/compilers/gcc/patches/default.nix:157]: Turns out gcc 13 includes the patch

[the remark this answers is not in the document]
Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/gcc/patches/default.nix:157]: Then why not just use

- https://github.com/NixOS/nixpkgs/pull/268097

**Comment by amjoseph-nixpkgs** [comment]
> ...and of course it's a mass rebuild because `optionalString false` returns an empty string, which is different from an empty LDFLAGS.

One of the many reasons to move towards `env = lib.optionalAttrs { LDFLAGS = ... }`...

File: pkgs/development/compilers/gcc/patches/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (fgaz) [reply pkgs/development/compilers/gcc/patches/default.nix:157]: Yes that's perfect!

I thought that upgrade was still far off, good to know that's not the case!

**Comment by fgaz** [comment]
Closing due to https://github.com/NixOS/nixpkgs/pull/268067#discussion_r1398321911.

I'll apply the workaround to cmake in another pr so risc-v is not blocked for another 1-2 staging cycles, then once #145377 is in I'll remove all workarounds.

**Comment by NickCao** [comment]
Workarounds removed in https://github.com/NixOS/nixpkgs/pull/274827

**Pull request closed without merging**